### PR TITLE
ARTEMIS-1740: Add support for regex based certificate authentication

### PR DESCRIPTION
**Pull request by LionelCons**
This adds the possibility to have an optional properties file containing regular expressions to match against the DN.

### CODE REVIEW

**Comment by clebertsuconic** [comment]
@jbertram can you look into this one? you have more miles on security than I do. :)

**Comment by jbertram** [comment]
This looks good except for a few things:

- I would rather not have an additional file for the regex properties. I think it would be better to instead denote the regex in the normal user properties file using the traditional forward slashes (as described in the JIRA).
- There's no documentation. The new functionality should be documented in docs/user-manual/en/security.md.
- I'd like to see an integration test added to {{org.apache.activemq.artemis.tests.integration.security.SecurityTest}}. There's several tests in there already which test cert-based login.

File: artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/ReloadableProperties.java [hunk 1]
```diff
@@ -95,6 +99,21 @@ public synchronized ReloadableProperties obtained() {
       return invertedValueProps;
    }
 
+   public synchronized Map<String, Pattern> regexpPropertiesMap() {
```
Comment (franz1981) [inline artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/ReloadableProperties.java]: You could use a [Supplier::memoize](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier) to allow a thread-safe lazy initialization without having that method synchronized even when you just need to get `regexpProps`

File: artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/TextFileCertificateLoginModule.java [hunk 2]
```diff
@@ -71,8 +79,12 @@ protected String getUserNameForCertificates(final X509Certificate[] certs) throw
       if (certs == null) {
          throw new LoginException("Client certificates not found. Cannot authenticate.");
       }
-
-      return usersByDn.get(getDistinguishedName(certs));
+      String dn = getDistinguishedName(certs);
+      String name = usersByDn.get(dn);
+      if (name == null && regexpByUser != null) {
+         name = getUserByRegexp(dn);
```
Comment (franz1981) [inline artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/TextFileCertificateLoginModule.java]: `getUserByRegexp` is synchronized but there are uses of `usersByDn` and `regexpByUser`, like these ones, that are not: what is the reason to have `String getUserByRegexp(String dn)` synchronized?

**Comment by LionelCons** [comment]
The intent of  `synchronized` is to protect the modification of `usersByDn`.

**Comment by franz1981** [comment]
@LionelCons 

And it makes sense, but there are parts like:
`String name = usersByDn.get(dn);`
that are not synchronized, hence it won't be thread-safe.
My advice is to use a lazy initialization using a `Suppliers::memoize` or similar construct and always use `Supplier::get` to access the variable: that would allow to have always a thread-safe access to an unmodifiable resource, lazy initialized.

**Comment by LionelCons** [comment]
@franz1981 your comment goes beyond my modifications since the existing code already uses the `get` method (in `getUserNameForCertificates` and `getUserRoles`, at least).

File: artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/TextFileCertificateLoginModule.java [hunk 3]
```diff
@@ -71,8 +79,12 @@ protected String getUserNameForCertificates(final X509Certificate[] certs) throw
       if (certs == null) {
          throw new LoginException("Client certificates not found. Cannot authenticate.");
       }
-
-      return usersByDn.get(getDistinguishedName(certs));
+      String dn = getDistinguishedName(certs);
+      String name = usersByDn.get(dn);
```
Comment (franz1981) [inline artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/TextFileCertificateLoginModule.java]: Hence this `userByDb` not synchronized access no longer exists in the new commit?
With github and multiple commits is not simple to do reviews :P

**Comment by LionelCons** [comment]
The final code can be seen [here](https://github.com/LionelCons/activemq-artemis/blob/9ead970395975a28846de4ba237eb78bc728c8fb/artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/TextFileCertificateLoginModule.java).

Just like in the original code, both `getUserNameForCertificates` and `getUserRoles` still use `get`.

**Comment by jbertram** [comment]
Thanks for the changes, @LionelCons. Can you squash your commits?

**Comment by LionelCons** [comment]
@jbertram can you squash the commits when merging?

**Comment by clebertsuconic** [comment]
@LionelCons I can.. if this is ready to be merged.

**Comment by clebertsuconic** [comment]
done

**Pull request closed without merging**